quinn-proto/src/connection/paths.rs

```diff
@@ -163,3 +170,453 @@ impl RttEstimator {
+            probe_packet_number: None,
+        }
+    }
+
+    fn new(initial_plpmtu: u16, config: PlpmtudConfig) -> Self {
+        debug_assert!(
+            initial_plpmtu >= BASE_UDP_PAYLOAD_SIZE,
+            "initial_max_udp_payload_size must be at least {BASE_UDP_PAYLOAD_SIZE}"
+        );
+
+        Self {
+            next_mtud_activation: None,
+            current_mtu: initial_plpmtu,
+            probing_udp_payload_size: initial_plpmtu,
+            black_hole_counter: 0,
+            largest_acked_mtu_sized_packet: None,
+            probe_packet_number: None,
+            lost_probes: 0,
+            search_lower_bound: initial_plpmtu,
+            search_upper_bound: config.max_udp_payload_size_upper_bound.max(initial_plpmtu),
+            phase: Phase::Searching,
+            config: Some(config),
+        }
+    }
+
+    fn from_previous(prev: &MtuDiscovery) -> Self {
+        if let Some(config) = &prev.config {
+            MtuDiscovery {
+                phase: prev.phase,
+                ..MtuDiscovery::new(prev.current_mtu, config.clone())
+            }
+        } else {
+            MtuDiscovery::disabled(prev.current_mtu)
+        }
+    }
+
+    /// Returns the amount of bytes that should be sent as an MTU probe, if any
+    pub fn poll_transmit(&mut self, now: Instant, next_packet_number: u64) -> Option<u16> {
+        if self.phase == Phase::Complete && self.next_mtud_activation.map_or(false, |i| i <= now) {
+            self.next_mtud_activation = None;
+            self.phase = Phase::Searching;
+            self.search_lower_bound = self.current_mtu;
+            self.search_upper_bound = self
+                .config
+                .as_ref()
+                .unwrap()
+                .max_udp_payload_size_upper_bound
+                .max(self.search_lower_bound);
+            self.probing_udp_payload_size = self.current_mtu;
+        }
+
+        if self.phase == Phase::Complete || self.probe_packet_number.is_some() {
+            return None;
+        }
+
+        // Retransmit lost probes
+        if 0 < self.lost_probes && self.lost_probes < MAX_PROBE_RETRANSMITS {
+            self.probe_packet_number = Some(next_packet_number);
+            return Some(self.probing_udp_payload_size);
+        }
+
+        let last_probe_succeeded = self.lost_probes == 0;
+
+        // The probe is definitely lost (we reached the MAX_PROBE_RETRANSMITS threshold)
+        if !last_probe_succeeded {
+            self.lost_probes = 0;
+            self.probe_packet_number = None;
+        }
+
+        if let Some(probe_udp_payload_size) = self.next_mtu_to_probe(last_probe_succeeded) {
+            self.probe_packet_number = Some(next_packet_number);
+            self.probing_udp_payload_size = probe_udp_payload_size;
+            Some(probe_udp_payload_size)
+        } else {
+            self.phase = Phase::Complete;
+            self.next_mtud_activation = Some(now + self.config.as_ref().unwrap().interval);
```

Comment:
Should we randomize this a little to avoid storms if many clients start at the same time (e.g. due to a server restart)?

Reply:
The current behavior is to run MTUD right after the connection is established, and when that is done, to run it on an interval. This line of code is scheduling MTUD to run again after the user-configured interval has elapsed, so I assume it shouldn't interact with something like a server restart. Or am I missing something?

In case you were referring to randomizing the moment of the initial MTUD run, I'm not convinced that it would be beneficial. For one, I assume in most cases you want to run MTUD as early as possible. Also, even if MTUD is disabled / randomized, a server will probably still need additional mechanisms to protect itself from a storm of connections... So my hunch here is to leave it as is and let servers deal with the problem through other means (e.g. load balancing)

Reply:
Makes sense. Probably this should be mitigated by higher level application logic acting on connection initiation timing.